File: internal/grafana/grafana.go
// Package grafana mints short-lived, read-only Grafana Cloud tokens and pairs
// them with the static query endpoints/instance IDs needed to use them.
package grafana

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/dlddu/homelab-k3s-mcp/internal/version"
)

const (
	// defaultAPIBase is the Grafana Cloud API base, including the /api segment.
	// The token path (/v1/tokens) is appended to it, so GRAFANA_API_URL must
	// likewise point at the .../api base (not the bare host).
	defaultAPIBase    = "https://www.grafana.com/api"
	tokenTTL          = time.Hour
	httpClientTimeout = 10 * time.Second
)

// errKind separates a "not configured" failure from an apiserver error.
type errKind int

const (
	kindUnavailable errKind = iota
	kindAPI
)

// Error is the error type returned by Service.
type Error struct {
	kind errKind
	msg  string
}

func (e *Error) Error() string {
	switch e.kind {
	case kindUnavailable:
		return "grafana cloud unavailable: " + e.msg
	default:
		return "grafana cloud api error: " + e.msg
	}
}

func unavailable(msg string) *Error { return &Error{kind: kindUnavailable, msg: msg} }
func apiError(msg string) *Error    { return &Error{kind: kindAPI, msg: msg} }

// Credentials bundles a freshly minted short-lived token with the static
// Grafana Cloud query endpoints and their numeric instance IDs. Token is the
// shared HTTP Basic password for both the metrics and logs users.
type Credentials struct {
	Token       string
	ExpiresAt   string
	MetricsURL  string
	MetricsUser string
	LogsURL     string
	LogsUser    string
}

// Service mints short-lived read-only Grafana Cloud credentials. The access
// policy (and therefore the scope) and the one-hour TTL are fixed on the
// server, so the call takes no arguments.
type Service interface {
	CreateToken(ctx context.Context) (*Credentials, error)
}

// Unavailable is a Service that fails every call with the same reason.
type Unavailable struct {
	reason string
}

// NewUnavailable builds an Unavailable service with the given reason.
func NewUnavailable(reason string) *Unavailable {
	if reason == "" {
		reason = "grafana cloud credentials are not configured"
	}
	return &Unavailable{reason: reason}
}

// CreateToken always fails with the configured reason.
func (u *Unavailable) CreateToken(context.Context) (*Credentials, error) {
	return nil, unavailable(u.reason)
}

// Client is the live Grafana Cloud implementation of Service.
type Client struct {
	issuerToken  string
	readPolicyID string
	region       string
	apiBase      string
	metricsURL   string
	metricsUser  string
	logsURL      string
	logsUser     string
	userAgent    string
	http         *http.Client
}

// FromEnv builds a Client from the GRAFANA_* environment variables. It returns
// (nil, nil) when GRAFANA_ISSUER_TOKEN is unset, signalling that the Grafana
// integration is simply not configured. When the issuer token is set but a
// required companion variable is missing, it returns an error (a
// misconfiguration) rather than silently dropping it.
func FromEnv() (*Client, error) {
	issuerToken := os.Getenv("GRAFANA_ISSUER_TOKEN")
	if issuerToken == "" {
		return nil, nil
	}

	readPolicyID := os.Getenv("GRAFANA_READ_POLICY_ID")
	if readPolicyID == "" {
		return nil, fmt.Errorf("GRAFANA_READ_POLICY_ID is required when GRAFANA_ISSUER_TOKEN is set")
	}

	metricsURL := os.Getenv("GRAFANA_METRICS_URL")
	metricsUser := os.Getenv("GRAFANA_METRICS_USER")
	logsURL := os.Getenv("GRAFANA_LOGS_URL")
	logsUser := os.Getenv("GRAFANA_LOGS_USER")
	var missing []string
	for _, kv := range []struct{ name, val string }{
		{"GRAFANA_METRICS_URL", metricsURL},
		{"GRAFANA_METRICS_USER", metricsUser},
		{"GRAFANA_LOGS_URL", logsURL},
		{"GRAFANA_LOGS_USER", logsUser},
	} {
		if kv.val == "" {
			missing = append(missing, kv.name)
		}
	}
	if len(missing) > 0 {
		return nil, fmt.Errorf("%s required when GRAFANA_ISSUER_TOKEN is set", strings.Join(missing, ", "))
	}

	apiBase := os.Getenv("GRAFANA_API_URL")
	if apiBase == "" {
		apiBase = defaultAPIBase
	}

	return &Client{
		issuerToken:  issuerToken,
		readPolicyID: readPolicyID,
		region:       os.Getenv("GRAFANA_REGION"),
		apiBase:      strings.TrimRight(apiBase, "/"),
		metricsURL:   metricsURL,
		metricsUser:  metricsUser,
		logsURL:      logsURL,
		logsUser:     logsUser,
		userAgent:    version.Name + "/" + version.Version,
		http:         &http.Client{Timeout: httpClientTimeout},
	}, nil
}

// CreateToken mints a fresh token under the configured access policy that
// expires one hour from now and returns it alongside the static query config.
func (c *Client) CreateToken(ctx context.Context) (*Credentials, error) {
	name := tokenName()
	expiresAt := time.Now().Add(tokenTTL).UTC().Format(time.RFC3339)

	body := map[string]any{
		"accessPolicyId": c.readPolicyID,
		"name":           name,
		"displayName":    name,
		"expiresAt":      expiresAt,
	}
	bodyBytes, err := json.Marshal(body)
	if err != nil {
		return nil, apiError(fmt.Sprintf("encode request body: %v", err))
	}

	endpoint := c.apiBase + "/v1/tokens"
	if c.region != "" {
		endpoint += "?region=" + url.QueryEscape(c.region)
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, apiError(fmt.Sprintf("build request: %v", err))
	}
	req.Header.Set("Authorization", "Bearer "+c.issuerToken)
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", c.userAgent)

	resp, err := c.http.Do(req)
	if err != nil {
		return nil, apiError(fmt.Sprintf("post %s: %v", endpoint, err))
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, apiError(fmt.Sprintf("%s returned %s: %s", endpoint, resp.Status, string(respBody)))
	}

	var parsed struct {
		Token     string `json:"token"`
		ExpiresAt string `json:"expiresAt"`
	}
	if err := json.Unmarshal(respBody, &parsed); err != nil {
		return nil, apiError(fmt.Sprintf("parse token: %v", err))
	}
	if parsed.Token == "" {
		return nil, apiError("grafana response did not include a token")
	}
	return &Credentials{
		Token:       parsed.Token,
		ExpiresAt:   parsed.ExpiresAt,
		MetricsURL:  c.metricsURL,
		MetricsUser: c.metricsUser,
		LogsURL:     c.logsURL,
		LogsUser:    c.logsUser,
	}, nil
}

// tokenName builds a per-request unique token name. Grafana Cloud requires the
// name to be unique within an access policy, so a timestamp plus a random
// suffix avoids collisions across rapid or concurrent calls.
func tokenName() string {
	suffix := make([]byte, 4)
	if _, err := rand.Read(suffix); err != nil {
		return fmt.Sprintf("%s-%d", version.Name, time.Now().UnixNano())
	}
	return fmt.Sprintf("%s-%d-%s", version.Name, time.Now().Unix(), hex.EncodeToString(suffix))
}

File: src/grafana.rs
use std::fmt;
use std::time::Duration;

use async_trait::async_trait;
use chrono::SecondsFormat;
use serde::{Deserialize, Serialize};
use serde_json::json;

const DEFAULT_API_BASE: &str = "https://www.grafana.com/api";
const TOKEN_TTL_HOURS: i64 = 1;

#[derive(Debug)]
pub enum GrafanaError {
    Unavailable(String),
    Api(String),
}

impl fmt::Display for GrafanaError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            GrafanaError::Unavailable(msg) => write!(f, "grafana cloud unavailable: {msg}"),
            GrafanaError::Api(msg) => write!(f, "grafana cloud api error: {msg}"),
        }
    }
}

impl std::error::Error for GrafanaError {}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GrafanaCloudToken {
    pub token: String,
    #[serde(rename = "expiresAt")]
    pub expires_at: String,
    #[serde(default, rename = "accessPolicyId")]
    pub access_policy_id: String,
}

#[async_trait]
pub trait GrafanaCloudService: Send + Sync {
    async fn create_short_lived_token(&self) -> Result<GrafanaCloudToken, GrafanaError>;
}

pub struct UnavailableGrafanaCloud {
    reason: String,
}

impl UnavailableGrafanaCloud {
    pub fn new(reason: impl Into<String>) -> Self {
        Self {
            reason: reason.into(),
        }
    }
}

impl Default for UnavailableGrafanaCloud {
    fn default() -> Self {
        Self::new("grafana cloud credentials are not configured")
    }
}

#[async_trait]
impl GrafanaCloudService for UnavailableGrafanaCloud {
    async fn create_short_lived_token(&self) -> Result<GrafanaCloudToken, GrafanaError> {
        Err(GrafanaError::Unavailable(self.reason.clone()))
    }
}

pub struct GrafanaCloudClient {
    management_token: String,
    access_policy_id: String,
    region: String,
    api_base: String,
    user_agent: String,
    http: reqwest::Client,
}

impl GrafanaCloudClient {
    pub fn from_env() -> Result<Option<Self>, String> {
        let management_token = match std::env::var("GRAFANA_CLOUD_ACCESS_POLICY_TOKEN") {
            Ok(v) if !v.is_empty() => v,
            _ => return Ok(None),
        };

        let access_policy_id = std::env::var("GRAFANA_CLOUD_ACCESS_POLICY_ID")
            .ok()
            .filter(|s| !s.is_empty())
            .ok_or_else(|| {
                "GRAFANA_CLOUD_ACCESS_POLICY_ID is required when \
                 GRAFANA_CLOUD_ACCESS_POLICY_TOKEN is set"
                    .to_string()
            })?;

        let region = std::env::var("GRAFANA_CLOUD_REGION")
            .ok()
            .filter(|s| !s.is_empty())
            .ok_or_else(|| {
                "GRAFANA_CLOUD_REGION is required when GRAFANA_CLOUD_ACCESS_POLICY_TOKEN is set"
                    .to_string()
            })?;

        let api_base = std::env::var("GRAFANA_API_BASE_URL")
            .ok()
            .filter(|s| !s.is_empty())
            .unwrap_or_else(|| DEFAULT_API_BASE.to_string());

        let user_agent = format!("{}/{}", env!("CARGO_PKG_NAME"), env!("CARGO_PKG_VERSION"));

        let http = reqwest::Client::builder()
            .timeout(Duration::from_secs(10))
            .build()
            .map_err(|e| format!("build http client: {e}"))?;

        Ok(Some(Self {
            management_token,
            access_policy_id,
            region,
            api_base: api_base.trim_end_matches('/').to_string(),
            user_agent,
            http,
        }))
    }
}

#[async_trait]
impl GrafanaCloudService for GrafanaCloudClient {
    async fn create_short_lived_token(&self) -> Result<GrafanaCloudToken, GrafanaError> {
        let now = chrono::Utc::now();
        let expires_at = (now + chrono::Duration::hours(TOKEN_TTL_HOURS))
            .to_rfc3339_opts(SecondsFormat::Secs, true);
        // Grafana requires token names to be unique within (org, region); the
        // millisecond timestamp keeps repeated mints from colliding.
        let name = format!("{}-{}", env!("CARGO_PKG_NAME"), now.timestamp_millis());

        let url = format!("{}/v1/tokens?region={}", self.api_base, self.region);
        let body = json!({
            "accessPolicyId": self.access_policy_id,
            "name": name,
            "displayName": name,
            "expiresAt": expires_at,
        });

        let response = self
            .http
            .post(&url)
            .bearer_auth(&self.management_token)
            .header(reqwest::header::ACCEPT, "application/json")
            .header(reqwest::header::USER_AGENT, &self.user_agent)
            .json(&body)
            .send()
            .await
            .map_err(|e| GrafanaError::Api(format!("post {url}: {e}")))?;

        let status = response.status();
        if !status.is_success() {
            let text = response.text().await.unwrap_or_default();
            return Err(GrafanaError::Api(format!(
                "{url} returned {status}: {text}"
            )));
        }

        let mut token = response
            .json::<GrafanaCloudToken>()
            .await
            .map_err(|e| GrafanaError::Api(format!("parse grafana cloud token: {e}")))?;
        token.access_policy_id = self.access_policy_id.clone();
        Ok(token)
    }
}
